packages/agent/src/agent/http/index.ts

```diff
@@ -1228,50 +1235,73 @@ export class HttpAgent implements Agent {
+    this.#syncTimePromise =
+      this.#syncTimePromise ??
+      (async () => {
+        await this.#rootKeyGuard();
+        const CanisterStatus = await import('../../canisterStatus');
+        const callTime = Date.now();
+        try {
+          if (!canisterIdOverride) {
+            this.log.print(
+              'Syncing time with the IC. No canisterId provided, so falling back to ryjl3-tyaaa-aaaaa-aaaba-cai',
+            );
+          }
+          // Fall back with canisterId of the ICP Ledger
+          const canisterId = canisterIdOverride ?? Principal.from('ryjl3-tyaaa-aaaaa-aaaba-cai');
+
+          const anonymousAgent = HttpAgent.createSync({
+            identity: new AnonymousIdentity(),
+            host: this.host.toString(),
+            fetch: this.#fetch,
+            retryTimes: 0,
+            rootKey: this.rootKey ?? undefined,
+          });
+
+          const replicaTimes = await Promise.all(
+            Array(3)
+              .fill(null)
+              .map(async () => {
+                const status = await CanisterStatus.request({
+                  canisterId,
+                  agent: anonymousAgent,
+                  paths: ['time'],
+                });
+
+                const date = status.get('time');
+                if (date instanceof Date) {
+                  return date.getTime();
+                }
+              }, []),
+          );
+
+          const maxReplicaTime = replicaTimes.reduce<number>((max, current) => {
+            return typeof current === 'number' && current > max ? current : max;
+          }, 0);
+
+          if (maxReplicaTime > BigInt(0)) {
+            this.#timeDiffMsecs = Number(maxReplicaTime) - Number(callTime);
+            this.log.notify({
+              message: `Syncing time: offset of ${this.#timeDiffMsecs}`,
+              level: 'info',
+            });
+          }
+        } catch (error) {
+          const syncTimeError =
+            error instanceof AgentError
+              ? error
+              : UnknownError.fromCode(new UnexpectedErrorCode(error));
+          this.log.error('Caught exception while attempting to sync time', syncTimeError);
+
+          throw syncTimeError;
+        }
+      })();
 
-      const replicaTime = status.get('time');
-      if (replicaTime) {
-        this.#timeDiffMsecs = Number(replicaTime as bigint) - Number(callTime);
-        this.log.notify({
-          message: `Syncing time: offset of ${this.#timeDiffMsecs}`,
-          level: 'info',
-        });
-      }
-    } catch (error) {
-      let syncTimeError: AgentError;
-      if (error instanceof AgentError) {
-        syncTimeError = error;
-      } else {
-        syncTimeError = UnknownError.fromCode(new UnexpectedErrorCode(error));
-      }
-      this.log.error('Caught exception while attempting to sync time', syncTimeError);
-    }
+    await this.#syncTimePromise;
+    this.#syncTimePromise = null;
```

Comment:
Should we reset the promise if it throws an error or not?

```suggestion
    await this.#syncTimePromise.finally(() => this.#syncTimePromise = null);
```

Same for `#rootKeyPromise`.

We should add tests to cover this case

Reply:
We can introduce tests in another PR